htmltools/_core.py

```diff
@@ -89,23 +90,38 @@ class MetadataNode:
 
 TagT = TypeVar("TagT", bound="Tag")
 
-# Types that can be passed in as attributes to tag functions.
-TagAttrArg = Union[str, float, bool, None]
+# Types that can be passed in as attributes to Tag functions. These values will be
+# converted to strings before being stored as tag attributes.
+TagAttrValue = Union[str, float, bool, None]
 
-# Types of objects that can be a child of a tag.
-TagChild = Union["Tagifiable", "Tag", MetadataNode, str]
+# For dictionaries of tag attributes (e.g., {"id": "foo"}), which can be passed as
+# unnamed arguments to Tag functions like ``div()``.
+TagAttrs = Dict[str, TagAttrValue]
 
-# Types that can be passed as args to TagList() and tag functions.
-TagChildArg = Union[
-    TagChild,
+# Types of objects that can be a node in a Tag tree. Equivalently, these are the valid
+# elements of a TagList. Note that this type represents the internal structure of items
+# in a TagList; the user-facing type is TagChild.
+TagNode = Union["Tagifiable", "Tag", MetadataNode, str]
+
+# Types of objects that can be passed as children to Tag functions like ``div()``. The
+# Tag functions and the TagList() constructor can accept these as unnamed arguments;
+# they will be flattened and normalized to TagNode objects.
+TagChild = Union[
+    TagNode,
     "TagList",
     float,
     None,
-    "dict[str, TagAttrArg]",  # i.e., tag attrbutes (e.g., {"id": "foo"})
-    Sequence["TagChildArg"],
+    Sequence["TagChild"],
 ]
 
 
+# These two types existed in htmltools 0.14.0 and earlier. They are here so that
+# existing versions of Shiny will be able to load, but users of those existing packages
+# will see type errors, which should encourage them to upgrade Shiny.
+TagChildArg = Never
+TagAttrArg = Never
```

Comment:
Just curious, would it be possible to throw a warning if they do actually get imported?

Reply:
Not that I know of. I saw this on SO, but it throws warnings a bit too easily. (https://stackoverflow.com/a/60486106/412655)

```py
class TagChildArg:
   raise DeprecationWarning("text here")
```

The problem is that it raises the warning even if you just do `from htmltools import div`.